faunadb-scala/src/main/scala/faunadb/streaming/BodyValueFlowProcessor.scala

```diff
@@ -0,0 +1,83 @@
+package faunadb.streaming
+
+import java.nio.ByteBuffer
+import java.util
+import java.util.concurrent.{Flow, SubmissionPublisher}
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.{JsonMappingException, ObjectMapper}
+import faunadb.errors.StreamingException
+import faunadb.values.{StringV, VSuccess, Value}
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+
+class BodyValueFlowProcessor extends SubmissionPublisher[Value] with Flow.Processor[util.List[ByteBuffer], Value] {
+  private val log = LoggerFactory.getLogger(getClass)
+  private val json = new ObjectMapper
+  private var subscription: java.util.concurrent.Flow.Subscription = null
+  private var subscriber: Flow.Subscriber[_ >: Value] = null
+
+  // We do not request data from the publisher until we have one subscriber
+  override def subscribe(subscriber: Flow.Subscriber[_ >: Value]): Unit = {
+    if (this.subscriber == null) {
+      this.subscriber = subscriber
+      super.subscribe(subscriber)
+      requestOne()
+    } else
+      throw new IllegalStateException("BodyValueFlowProcessor can have only one subscriber")
+  }
+
+  override def onSubscribe(subscription: Flow.Subscription): Unit =
+    this.subscription = subscription
+
+  override def onNext(item: util.List[ByteBuffer]): Unit = {
+    import java.nio.charset.StandardCharsets
+    val text = item.iterator().asScala.map(StandardCharsets.US_ASCII.decode(_).toString).mkString("")
+    try {
+      val jsonNode = json.readTree(text)
+      val value = json.treeToValue[Value](jsonNode, classOf[Value])
+      isUnrecoverableError(value) match {
+        case None => submit(value)
+        case Some(unrecoverableError) =>
+          subscriber.onError(unrecoverableError) // notify subscriber stream
+          subscription.cancel() // cancel subscription on the request body
+      }
+    } catch {
+      case e: JsonProcessingException => log.error(s"could not parse JSON event $text", e)
+      case e: JsonMappingException => log.error(s"could not parse JSON event $text", e)
+      case e: Throwable => log.error(s"could not parse event $text", e)
+    }
+
+    requestOne()
+  }
+
+  private def isUnrecoverableError(event: Value): Option[Throwable] = {
+    (event("event"), event("data", "code")) match {
+      case (VSuccess(StringV("error"), _), VSuccess(StringV("permission denied"), _)) =>
+        val description = event("data", "description").map(_.toString).getOrElse("no description available")
+        Some(StreamingException(s"code: permission denied, description:$description", null))
```

Comment:
I'm wondering if (from a user perspective) it would be practical to deal with a new `Exception` here or if we could just return the `PermissionDeniedException` instead. 🤔

Reply:
I thought about this too 👍  
However my problem was that the whole exception family to which `PermissionDeniedException` belongs is centered around wrapping an `HttpResponse` which I do not have here.
It might require a larger refactoring.

WDYT?

Reply:
Mhm, I think we should try to reuse `PermissionDeniedException` here if possible. 

The `PermissionDeniedException` has an overloaded constructor that accepts just a `String` value. Maybe we can use that one for creating it here? :)

https://github.com/fauna/faunadb-jvm/blob/88092e74be0d74ec847fa9bea6299cf1aac98f54/faunadb-scala/src/main/scala/faunadb/errors/FaunaException.scala#L65

In case we cannot and this requires a more complicated refactor, let's push it back and see if we have time to do it or not later after the other changes before merging the PR.